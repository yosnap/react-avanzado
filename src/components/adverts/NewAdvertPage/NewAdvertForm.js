import T from 'prop-types';

import { InputFile } from '../../common';
import { Form, FormConsumer, Input } from '../../form';
import SelectTags from '../SelectTags';

const validName = ({ name }) => name;
const validPrice = ({ price }) =>
  !Number.isNaN(price) && Number.isFinite(price) && price >= 0;
const validTags = ({ tags }) => !!tags.length;

function NewAdvertForm({ onSubmit }) {
  return (
    <div className='form-box'>
      <Form
        onSubmit={onSubmit}
        initialValue={{ name: '', sale: true, price: 0, tags: [], photo: null }}
      >
        <label>
          Name
          <Input name="name" />
        </label>
        <label>
          Sell
          <Input type="checkbox" name="sale" />
        </label>
        <label>
          Price
          <Input type="number" name="price" />
        </label>
        <label>
          Tags
          <Input component={SelectTags} name="tags" />
        </label>
        <div className='img-input'>
        <Input component={InputFile} name="photo" data-testid="photo-input" />
        </div>
        <FormConsumer>
          {({ validate }) => (
            <button disabled={!validate(validName, validPrice, validTags)}>
              Save
            </button>
          )}
        </FormConsumer>
      </Form>
    </div>
  );
}

NewAdvertForm.propTypes = {
  onSubmit: T.func.isRequired,
};

export default NewAdvertForm;
