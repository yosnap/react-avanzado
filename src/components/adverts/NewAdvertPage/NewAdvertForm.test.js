import { fireEvent, render, waitFor } from '@testing-library/react';

import { Provider } from 'react-redux';
import NewAdvertForm from './NewAdvertForm';

const store = {
  getState: () => ({ tags: ['motor', 'mobile', 'lifestyle', 'work'] }),
  subscribe: () => {},
  dispatch: () => {},
};

const renderNewAdvertForm = props => (
  <Provider store={store}>
    <NewAdvertForm {...props} />;
  </Provider>
);

describe('NewAdvertForm', () => {
  test('snapshot', () => {
    const { container } = render(renderNewAdvertForm({ onSubmit: () => {} }));
    expect(container).toMatchSnapshot();
  });

  test('should call onSubmit with the new advert', async () => {
    const onSubmit = jest.fn();
    const name = 'Coche';
    const price = 1200;
    const tags = ['motor'];
    const photo = new File(['photo'], 'photo.png', {
      type: 'image/png',
    });

    const {
      getByLabelText,
      getByRole,
      getByDisplayValue,
      getByAltText,
      getByTestId,
    } = render(renderNewAdvertForm({ onSubmit }));

    const nameField = getByLabelText(/name/i);
    const priceField = getByLabelText(/price/i);
    const sellField = getByLabelText(/sell/i);
    const motorField = getByDisplayValue(/motor/i);
    const photoField = getByTestId('photo-input');
    const submitButton = getByRole('button');

    expect(submitButton).toBeDisabled();

    fireEvent.change(nameField, { target: { value: name } });
    fireEvent.click(sellField);
    fireEvent.change(priceField, { target: { value: price } });
    fireEvent.click(motorField);
    fireEvent.change(photoField, { target: { files: [photo] } });

    await waitFor(() => getByAltText(photo.name));

    expect(submitButton).not.toBeDisabled();

    fireEvent.click(submitButton);
    expect(onSubmit).toHaveBeenCalledWith({
      name,
      price,
      tags,
      photo,
      sale: false,
    });
  });
});
