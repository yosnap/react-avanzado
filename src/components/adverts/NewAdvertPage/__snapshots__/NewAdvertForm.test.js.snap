// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewAdvertForm snapshot 1`] = `
<div>
  <form>
    <label>
      Name
      <input
        name="name"
        value=""
      />
    </label>
    <label>
      Sell
      <input
        checked=""
        name="sale"
        type="checkbox"
      />
    </label>
    <label>
      Price
      <input
        name="price"
        type="number"
        value="0"
      />
    </label>
    <label>
      Tags
      <div>
        <label>
          <input
            name="tags"
            type="checkbox"
            value="motor"
          />
          motor
        </label>
        <label>
          <input
            name="tags"
            type="checkbox"
            value="mobile"
          />
          mobile
        </label>
        <label>
          <input
            name="tags"
            type="checkbox"
            value="lifestyle"
          />
          lifestyle
        </label>
        <label>
          <input
            name="tags"
            type="checkbox"
            value="work"
          />
          work
        </label>
      </div>
    </label>
    <input
      data-testid="photo-input"
      name="photo"
      style="display: none;"
      type="file"
    />
    <img
      alt="placeholder.png"
      height="200"
      src="placeholder.png"
      style="object-fit: contain;"
      width="200"
    />
    <button
      disabled=""
    >
      Save
    </button>
  </form>
  ;
</div>
`;
